File: VidoX_Google/app/src/main/java/com/ashbash/vidoxproject/util/URLNormalizer.kt
package com.ashbash.vidoxproject.util

import com.ashbash.vidoxproject.models.VideoPlatform
import java.net.URI
import java.net.URL
import java.net.URLDecoder
import java.net.URLEncoder

/** Cleans pasted URLs so extraction is more reliable across platforms. */
object URLNormalizer {
    private val stripQueryKeys = setOf(
        "fbclid", "fb_action_ids", "fb_action_types", "fb_source",
        "mibextid", "rdid", "share_url", "ref", "refsrc", "_rdr",
        "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content",
        "si", "feature", "pp", "s", "refid", "__tn__", "sfnsn"
    )

    fun url(from: String): URL? {
        var trimmed = from.trim()
        if (trimmed.isEmpty()) return null

        if ((trimmed.startsWith("\"") && trimmed.endsWith("\"")) ||
            (trimmed.startsWith("'") && trimmed.endsWith("'")) ||
            (trimmed.startsWith("<") && trimmed.endsWith(">"))
        ) {
            trimmed = trimmed.substring(1, trimmed.length - 1).trim()
        }

        if (!trimmed.contains("://")) {
            trimmed = "https://$trimmed"
        }

        return try {
            val uri = URI(trimmed)
            val scheme = uri.scheme?.lowercase() ?: return null
            if (scheme != "http" && scheme != "https") return null

            var host = uri.host?.lowercase() ?: return null
            host = canonicalizeHost(host)

            val path = uri.path ?: ""
            val query = uri.rawQuery
            val rebuilt = buildString {
                append(scheme).append("://").append(host)
                append(if (path.isEmpty()) "" else path)
                if (!query.isNullOrEmpty()) {
                    val filtered = filterQuery(query, host, path)
                    if (filtered.isNotEmpty()) append('?').append(filtered)
                }
            }
            URL(rebuilt)
        } catch (_: Exception) {
            null
        }
    }

    private fun canonicalizeHost(host: String): String = when (host) {
        "m.facebook.com", "mbasic.facebook.com", "web.facebook.com", "mobile.facebook.com" ->
            "www.facebook.com"
        "fb.com", "www.fb.com" -> "www.facebook.com"
        "www.fb.watch" -> "fb.watch"
        else -> host
    }

    private fun filterQuery(rawQuery: String, host: String, path: String): String {
        val isFacebook = VideoPlatform.detect("https://$host$path") == VideoPlatform.FACEBOOK
        val keepFacebook = setOf("v", "story_fbid", "id", "video_id")
        return rawQuery.split("&").mapNotNull { pair ->
            val parts = pair.split("=", limit = 2)
            val name = parts[0].lowercase()
            if (isFacebook) {
                if (name in keepFacebook || name !in stripQueryKeys) pair else null
            } else {
                if (name in stripQueryKeys) null else pair
            }
        }.joinToString("&")
    }

    fun facebookVideoID(from: URL): String? {
        val query = from.query
        if (!query.isNullOrEmpty()) {
            for (key in listOf("v", "video_id", "story_fbid")) {
                val value = query.split("&")
                    .map { it.split("=", limit = 2) }
                    .firstOrNull { it[0] == key }
                    ?.getOrNull(1)
                    ?.let { URLDecoder.decode(it, "UTF-8") }
                if (value != null && isDigits(value) && value.length >= 5) return value
            }
        }

        val parts = from.path.split('/').filter { it.isNotEmpty() }
        if (parts.isEmpty()) return null

        val marker = parts.indexOfFirst { it == "reel" || it == "videos" || it == "watch" }
        if (marker >= 0 && marker + 1 < parts.size) {
            val candidate = parts[marker + 1].substringBefore('?')
            if (isDigits(candidate) && candidate.length >= 5) return candidate
        }

        if (parts.size >= 3 && parts[1] == "videos" && isDigits(parts[2]) && parts[2].length >= 5) {
            return parts[2]
        }
        return null
    }

    private fun isDigits(value: String): Boolean =
        value.isNotEmpty() && value.all { it.isDigit() }

    fun encodeQuery(value: String): String =
        URLEncoder.encode(value, "UTF-8")
}

File: VidoX_Apple/VidoXProject/Utilities/URLNormalizer.swift
import Foundation

/// Cleans pasted URLs so extraction is more reliable across platforms.
enum URLNormalizer {
    /// Tracking / share noise that breaks some extractors when left on the URL.
    private static let stripQueryKeys: Set<String> = [
        "fbclid", "fb_action_ids", "fb_action_types", "fb_source",
        "mibextid", "rdid", "share_url", "ref", "refsrc", "_rdr",
        "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content",
        "si", "feature", "pp", "s", "refid", "__tn__", "sfnsn"
    ]

    /// Trims whitespace/quotes, adds `https://` when missing, and rejects non-http(s) schemes.
    static func url(from raw: String) -> URL? {
        var trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        // Strip wrapping quotes copied from browsers or chat apps.
        if (trimmed.hasPrefix("\"") && trimmed.hasSuffix("\""))
            || (trimmed.hasPrefix("'") && trimmed.hasSuffix("'"))
            || (trimmed.hasPrefix("<") && trimmed.hasSuffix(">")) {
            trimmed = String(trimmed.dropFirst().dropLast())
            trimmed = trimmed.trimmingCharacters(in: .whitespacesAndNewlines)
        }

        if !trimmed.contains("://") {
            trimmed = "https://\(trimmed)"
        }

        guard var components = URLComponents(string: trimmed) else { return nil }
        guard let scheme = components.scheme?.lowercased(), scheme == "http" || scheme == "https" else {
            return nil
        }
        components.scheme = scheme

        if let host = components.host?.lowercased() {
            components.host = canonicalizeHost(host)
        }

        if VideoPlatform.detect(from: components.url?.absoluteString ?? trimmed) == .facebook {
            components = normalizeFacebookComponents(components)
        } else if VideoPlatform.detect(from: components.url?.absoluteString ?? trimmed) == .youtube {
            components = normalizeYouTubeComponents(components)
        } else if let items = components.queryItems, !items.isEmpty {
            components.queryItems = items.filter { item in
                !stripQueryKeys.contains(item.name.lowercased())
            }
            if components.queryItems?.isEmpty == true {
                components.queryItems = nil
            }
        }

        return components.url
    }

    /// Rewrites kkinstagram (and similar) hosts to instagram.com for extractors that
    /// only understand official Instagram URLs (yt-dlp).
    static func instagramCanonicalURL(from url: URL) -> URL {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let host = components.host?.lowercased(),
              host == "kkinstagram.com" || host.hasSuffix(".kkinstagram.com") else {
            return url
        }
        components.host = "www.instagram.com"
        return components.url ?? url
    }

    // MARK: - Host / Facebook helpers

    private static func canonicalizeHost(_ host: String) -> String {
        switch host {
        case "m.facebook.com", "mbasic.facebook.com", "web.facebook.com", "mobile.facebook.com":
            "www.facebook.com"
        case "fb.com", "www.fb.com":
            "www.facebook.com"
        case "www.fb.watch":
            "fb.watch"
        case "m.youtube.com", "music.youtube.com", "www.music.youtube.com",
             "youtube-nocookie.com", "www.youtube-nocookie.com":
            "www.youtube.com"
        default:
            host
        }
    }

    private static func normalizeFacebookComponents(_ components: URLComponents) -> URLComponents {
        var components = components
        let path = components.path

        // /reel/<id> and /videos/<id> are stable; drop tracking query noise.
        if let items = components.queryItems, !items.isEmpty {
            let kept = items.filter { item in
                let name = item.name.lowercased()
                if name == "v" || name == "story_fbid" || name == "id" || name == "video_id" {
                    return true
                }
                return !stripQueryKeys.contains(name)
            }
            components.queryItems = kept.isEmpty ? nil : kept
        }

        // Collapse accidental trailing slash variants on share links.
        if path.hasPrefix("/share/"), path.count > 1, path.hasSuffix("/") {
            components.path = String(path.dropLast())
        }

        return components
    }

    private static func normalizeYouTubeComponents(_ components: URLComponents) -> URLComponents {
        var components = components
        if let items = components.queryItems, !items.isEmpty {
            let kept = items.filter { item in
                let name = item.name.lowercased()
                if name == "v" || name == "vi" || name == "t" || name == "list" {
                    return true
                }
                return !stripQueryKeys.contains(name)
            }
            components.queryItems = kept.isEmpty ? nil : kept
        }
        return components
    }

    /// Best-effort numeric Facebook video id from common URL shapes.
    static func facebookVideoID(from url: URL) -> String? {
        if let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems {
            for key in ["v", "video_id", "story_fbid"] {
                if let value = items.first(where: { $0.name == key })?.value,
                   isDigits(value),
                   value.count >= 5 {
                    return value
                }
            }
        }

        let parts = url.path.split(separator: "/").map(String.init)
        guard !parts.isEmpty else { return nil }

        if let reelIndex = parts.firstIndex(where: { $0 == "reel" || $0 == "videos" || $0 == "watch" }),
           reelIndex + 1 < parts.count {
            let candidate = parts[reelIndex + 1]
                .split(separator: "?").first
                .map(String.init) ?? parts[reelIndex + 1]
            if isDigits(candidate), candidate.count >= 5 {
                return candidate
            }
        }

        // /username/videos/<id>
        if parts.count >= 3, parts[1] == "videos",
           isDigits(parts[2]),
           parts[2].count >= 5 {
            return parts[2]
        }

        return nil
    }

    private static func isDigits(_ value: String) -> Bool {
        !value.isEmpty && value.unicodeScalars.allSatisfy { CharacterSet.decimalDigits.contains($0) }
    }
}
